driver/src/orderbook/streamed/updating_orderbook.rs

```diff
@@ -7,177 +7,144 @@ use crate::{
-                // without a retry we would be stuck with an outdated orderbook forever.
-                // Sleep for one second, so that we have time to flush the logs.
-                thread::sleep(Duration::from_secs(1));
-                process::exit(1);
-            }
-        });
+impl UpdatingOrderbook {
+    /// Blocks on attempting to build the orderbook. After the initial build the orderbook is
+    /// updated every time `get_auction_data` is called.
+    pub fn new(contract: Arc<dyn StableXContract + Send + Sync>, web3: Web3) -> Result<Self> {
+        let result = Self {
+            contract,
+            web3: web3.clone(),
+            context: Mutex::new(Context {
+                orderbook: Orderbook::default(),
+                last_handled_block: 0,
+                block_timestamp_reader: CachedBlockTimestampReader::new(web3),
+            }),
+        };
 
-        Self {
-            orderbook,
-            orderbook_ready,
-            _exit_tx: exit_tx,
+        // Perform the initial update.
+        {
+            let mut context = result
+                .context
+                .lock()
+                .map_err(|err| anyhow!("poison error: {}", err))?;
+            futures::executor::block_on(result.update_with_events(&mut context))?;
         }
-    }
-}
 
-impl StableXOrderBookReading for UpdatingOrderbook {
-    fn get_auction_data(&self, batch_id_to_solve: U256) -> Result<(AccountState, Vec<Order>)> {
-        ensure!(
-            self.orderbook_ready.load(Ordering::SeqCst),
-            "orderbook not yet ready"
-        );
-        self.orderbook
-            .lock()
-            .map_err(|err| anyhow!("poison error: {}", err))?
-            .get_auction_data(batch_id_to_solve)
+        Ok(result)
     }
-}
 
-/// Update the orderbook with events from the stream forever or until exit_indicator is dropped.
-///
-/// Returns Ok when exit_indicator is dropped.
-/// Returns Err if the stream ends.
-async fn update_with_events_forever(
-    web3: Web3,
-    contract: Arc<dyn StableXContract>,
-    orderbook: Arc<Mutex<Orderbook>>,
-    orderbook_ready: Arc<AtomicBool>,
-    mut block_timestamp_reader: CachedBlockTimestampReader<Web3>,
-    exit_indicator: oneshot::Receiver<()>,
-) -> Result<()> {
-    const POLL_INTERVALL: Duration = Duration::from_secs(15);
-    const BLOCK_RANGE: u64 = 25;
-
-    log::info!("Starting event based orderbook updating.");
-
-    // `select!` requires the futures to be fused and pinned.
-    let exit_indicator = exit_indicator.fuse();
-    pin_mut!(exit_indicator);
-
-    let mut last_handled_block = 0u64;
-    loop {
-        let current_block = web3.eth().block_number().compat().await?;
-        let from_block = last_handled_block.saturating_sub(BLOCK_RANGE);
+    async fn update_with_events(&self, context: &mut Context) -> Result<()> {
+        const BLOCK_RANGE: u64 = 25;
+        log::info!("Updating event based orderbook.");
```

Comment:
uber nit: If you think it adds value to know we are querying the current block we can leave this (however we can also see this from the HTTP transport logs. Otherwise I'd say let's keep our logging to a few high signal lines, but really not important.